Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -1081,13 +1098,422 @@ TextureCacheBase::TCacheEntry* TextureCacheBase::Load(const u32 stage)
+          entry->CreateReference(entry_to_update);
+          // Mark the texture update as used, as if it was loaded directly
+          entry->frameCount = FRAMECOUNT_INVALID;
+        }
+      }
+      else
+      {
+        // If the hash does not match, this EFB copy will not be used for anything, so remove it
+        iter.first = InvalidateTexture(iter.first);
+        continue;
+      }
+    }
+    ++iter.first;
+  }
+
+  return updated_entry;
+}
+
+TextureCacheBase::TCacheEntry*
+TextureCacheBase::CreateNormalTexture(const TextureLookupInformation& tex_info)
+{
+  // create the entry/texture
+  TextureConfig config;
+  config.width = tex_info.native_width;
+  config.height = tex_info.native_height;
+  config.levels = tex_info.computed_levels;
+  config.format = AbstractTextureFormat::RGBA8;
+  config.rendertarget = true;
+
+  TCacheEntry* entry = AllocateCacheEntry(config);
+  GFX_DEBUGGER_PAUSE_AT(NEXT_NEW_TEXTURE, true);
+
+  if (!entry)
+    return nullptr;
+
+  textures_by_address.emplace(tex_info.address, entry);
+  if (tex_info.texture_cache_safety_color_sample_size == 0 ||
+      std::max(tex_info.total_bytes, tex_info.palette_size) <=
+          (u32)tex_info.texture_cache_safety_color_sample_size * 8)
+  {
+    entry->textures_by_hash_iter = textures_by_hash.emplace(tex_info.full_hash, entry);
+  }
+
+  entry->SetGeneralParameters(tex_info.address, tex_info.total_bytes, tex_info.full_format, false);
+  entry->SetDimensions(tex_info.native_width, tex_info.native_height, tex_info.computed_levels);
+  entry->SetHashes(tex_info.base_hash, tex_info.full_hash);
+  entry->is_custom_tex = false;
+  entry->memory_stride = entry->BytesPerRow();
+  entry->SetNotCopy();
+
+  INCSTAT(stats.numTexturesUploaded);
+  SETSTAT(stats.numTexturesAlive, textures_by_address.size());
+
+  return entry;
+}
+
+void TextureCacheBase::LoadTextureFromMemory(TCacheEntry* entry_to_update,
+                                             const TextureLookupInformation& tex_info)
+{
+  // We can decode on the GPU if it is a supported format and the flag is enabled.
+  // Currently we don't decode RGBA8 textures from Tmem, as that would require copying from both
+  // banks, and if we're doing an copy we may as well just do the whole thing on the CPU, since
+  // there's no conversion between formats. In the future this could be extended with a separate
+  // shader, however.
+  bool decode_on_gpu = g_ActiveConfig.UseGPUTextureDecoding() &&
+                       g_texture_cache->SupportsGPUTextureDecode(tex_info.full_format.texfmt,
+                                                                 tex_info.full_format.tlutfmt) &&
+                       !(tex_info.from_tmem && tex_info.full_format.texfmt == TextureFormat::RGBA8);
+
+  LoadTextureLevelZeroFromMemory(entry_to_update, tex_info, decode_on_gpu);
+}
+
+void TextureCacheBase::LoadTextureLevelZeroFromMemory(TCacheEntry* entry_to_update,
```

Comment:
Is this function only used for XFB? Maybe "LoadXFBTexture" would be a better name?